drivers/i2c/i2c_emul.c

```diff
@@ -74,13 +83,96 @@ static int i2c_emul_get_config(const struct device *dev, uint32_t *dev_config)
 	return 0;
 }
 
+#ifdef CONFIG_I2C_TARGET
+static int i2c_emul_send_to_target(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs)
+{
+	struct i2c_emul_data *data = dev->data;
+	const struct i2c_target_callbacks *callbacks = data->target_cfg->callbacks;
+
+	for (uint8_t i = 0; i < num_msgs; ++i) {
+		LOG_DBG("    msgs[%u].flags? 0x%02x", i, msgs[i].flags);
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+			for (uint32_t j = 0; j < msgs[i].len; ++j) {
+				int rc = 0;
+
+				if (j == 0) {
+					LOG_DBG("    Calling read_requested with data %p",
+						(void *)&msgs[i].buf[j]);
+					rc = callbacks->read_requested(data->target_cfg,
+								       &msgs[i].buf[j]);
+				} else {
+					LOG_DBG("    Calling read_processed with data %p",
+						(void *)&msgs[i].buf[j]);
+					rc = callbacks->read_processed(data->target_cfg,
+								       &msgs[i].buf[j]);
+				}
+				if (rc != 0) {
+					return rc;
+				}
+			}
+		} else {
+			for (uint32_t j = 0; j < msgs[i].len; ++j) {
+				int rc = 0;
+
+				if (j == 0) {
+					LOG_DBG("    Calling write_requested");
+					rc = callbacks->write_requested(data->target_cfg);
+				}
+				if (rc != 0) {
+					return rc;
+				}
+				LOG_DBG("    Calling write_received with data 0x%02x",
+					msgs[i].buf[j]);
+				rc = callbacks->write_received(data->target_cfg, msgs[i].buf[j]);
+				if (rc != 0) {
+					return rc;
+				}
+			}
+		}
+		if ((msgs[i].flags & I2C_MSG_STOP) == I2C_MSG_STOP) {
+			int rc = callbacks->stop(data->target_cfg);
+
+			if (rc != 0) {
+				return rc;
+			}
+		}
+	}
+	return 0;
+}
+#endif /* CONFIG_I2C_TARGET*/
+
 static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
 			     uint16_t addr)
 {
+	const struct i2c_emul_config *conf = dev->config;
 	struct i2c_emul *emul;
 	const struct i2c_emul_api *api;
 	int ret;
 
+	LOG_DBG("%s(dev=%p, addr=0x%02x)", __func__, (void *)dev, addr);
+#ifdef CONFIG_I2C_TARGET
+	struct i2c_emul_data *data = dev->data;
+
+	LOG_DBG("    has_target_cfg? %d", data->target_cfg != NULL);
+	if (data->target_cfg != NULL) {
```

Comment:
Please add some comments that describe the 2 new use cases.  I2C transfers sent to an I2C target, and I2C transfers forwarded from a controller to target.

Reply:
Done